numba/npyufunc/_internal.c

```diff
@@ -104,6 +105,557 @@ PyTypeObject PyUFuncCleaner_Type = {
+{
+    return ufunc_dispatch.ufunc_reduce((PyObject*)self->ufunc, args, kws);
+}
+
+static PyObject *
+dufunc_accumulate(PyDUFuncObject * self, PyObject * args, PyObject *kws)
+{
+    return ufunc_dispatch.ufunc_accumulate((PyObject*)self->ufunc, args, kws);
+}
+
+static PyObject *
+dufunc_reduceat(PyDUFuncObject * self, PyObject * args, PyObject *kws)
+{
+    return ufunc_dispatch.ufunc_reduceat((PyObject*)self->ufunc, args, kws);
+}
+
+static PyObject *
+dufunc_outer(PyDUFuncObject * self, PyObject * args, PyObject *kws)
+{
+    return ufunc_dispatch.ufunc_outer((PyObject*)self->ufunc, args, kws);
+}
+
+static PyObject *
+dufunc_at(PyDUFuncObject * self, PyObject * args)
+{
+    return ufunc_dispatch.ufunc_at((PyObject*)self->ufunc, args);
+}
+
+static PyObject *
+dufunc__compile_for_args(PyDUFuncObject * self, PyObject * args,
+                        PyObject * kws)
+{
+    PyErr_SetString(PyExc_NotImplementedError,
+                    "Abstract method _DUFunc._compile_for_args() called!");
+    return NULL;
+}
+
+static int *
+_build_arg_types_array(PyObject * type_list, Py_ssize_t nargs)
+{
+    int *arg_types_array=NULL;
+    Py_ssize_t idx, arg_types_size = PyList_Size(type_list);
+
+    if (arg_types_size != nargs) {
+        PyErr_SetString(
+            PyExc_ValueError,
+            "argument type list size does not equal ufunc argument count");
+        return NULL;
+    }
+    arg_types_array = PyArray_malloc(sizeof(int) * nargs);
+    if (!arg_types_array) {
+        PyErr_NoMemory();
+        return NULL;
+    }
+    for (idx = 0; idx < nargs; idx++) {
+        arg_types_array[idx] = (int)PyLong_AsLong(PyList_GET_ITEM(type_list,
+                                                                  idx));
+    }
+    if (PyErr_Occurred()) {
+        PyArray_free(arg_types_array);
+        arg_types_array = NULL;
+    }
+    return arg_types_array;
+}
+
+static PyObject *
+dufunc__add_loop(PyDUFuncObject * self, PyObject * args)
+{
+    PyUFuncObject * ufunc=self->ufunc;
+    unsigned long loop_ptr=0, data_ptr=0;
+    int idx=-1, usertype=NPY_VOID;
+    int *arg_types_arr=NULL;
+    PyObject *arg_types=NULL;
+    PyUFuncGenericFunction old_func=NULL;
+
+    if (self->frozen) {
+        PyErr_SetString(PyExc_ValueError,
+                        "_DUFunc._add_loop() called for frozen dufunc");
+        return NULL;
+    }
+
+    if (!PyArg_ParseTuple(args, "kO!|k", &loop_ptr, &PyList_Type, &arg_types,
+                          &data_ptr)) {
+        return NULL;
+    }
+
+    arg_types_arr = _build_arg_types_array(arg_types, (Py_ssize_t)ufunc->nargs);
+    if (!arg_types_arr) goto _dufunc__add_loop_fail;
+    for (idx = 0; idx < ufunc->nargs; idx++) {
+        if (arg_types_arr[idx] >= NPY_USERDEF) {
+            usertype = arg_types_arr[idx];
+        }
+    }
+
+    if (usertype != NPY_VOID) {
+        if (PyUFunc_RegisterLoopForType(ufunc, usertype,
+                                        (PyUFuncGenericFunction)loop_ptr,
```

Comment:
I don't understand that comment :-) `unsigned long` is simply too small to contain a pointer on some architectures (e.g. 64-bit Windows, but anything with a LLP model rather than LP)